### PR TITLE
ConsolidatedDimensionCoords Transform -- cleaned up version of PR #556

### PR DESCRIPTION
**Pull request by norlandrhagen**
Reviving from a clean slate https://github.com/norlandrhagen/pangeo-forge-recipes/pull/new/consolidate_dimension_coords. This PR only intends to work on this issue: https://github.com/pangeo-forge/pangeo-forge-recipes/pull/556.

### CODE REVIEW

**Comment by norlandrhagen** [comment]
Unit and end2end tests added! Big thanks to @jbusecke for help here. Both of us went back to Zarr101 class to figure out how to create a dataset without Xarray.

**Comment by jbusecke** [comment]
Big thanks back! This was really fun, and I am excited to see this moving.

File: tests/test_end_to_end.py
Comment (ranchodeluxe) [inline tests/test_end_to_end.py:192]: do you mind seeing if this still works if we use `target_root=tmp_target` b/c that actually returns a `FSSpecTarget`?

File: pangeo_forge_recipes/rechunking.py
Comment (ranchodeluxe) [inline pangeo_forge_recipes/rechunking.py:246]: I imagine since this is a `zarr.Group` this iteration isn't very expensive b/c it's on summary data and never really that large?

**Review by ranchodeluxe — APPROVED** [review]
Makes sense to me (but I don't know much 😆). Like the tests. Left a couple non-blocking comments

File: pangeo_forge_recipes/rechunking.py
Comment (norlandrhagen) [reply pangeo_forge_recipes/rechunking.py:246]: Ehh, I think you're right? Since it's just scanning across vars in groups and looking at the .attrs.

File: tests/test_end_to_end.py
Comment (norlandrhagen) [reply tests/test_end_to_end.py:192]: Great call! I'll give it a test.

Comment (norlandrhagen) [reply tests/test_end_to_end.py:192]: Gotta figure out a way to append the path to an `FSSpecTarget`. 
This fails: `store = zarr.open(os.path.join(tmp_target + "subpath"))`

Comment (ranchodeluxe) [reply tests/test_end_to_end.py:192]: haha, if it doesn't just work then I'll deal with it in my PR so don't worry 😆

**Review by jbusecke — COMMENTED** [review]
Thanks @norlandrhagen. Ill try this in the wild with CMIP6 in a few. 

You can probably delete the TODO comments [here](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/6c39134e22f372390e4e609df26631b2116e2d55/pangeo_forge_recipes/transforms.py#L391-L393) too.

**Comment by jbusecke** [comment]
FYI: https://github.com/leap-stc/cmip6-leap-feedstock/pull/88

**Comment by jbusecke** [comment]
⚠️ This failed with all my test case cmip files with an error like this: 
```
Error message from worker: Traceback (most recent call last):
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/conventions.py", line 436, in decode_cf_variables
    new_vars[k] = decode_cf_variable(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/conventions.py", line 287, in decode_cf_variable
    var = times.CFDatetimeCoder(use_cftime=use_cftime).decode(var, name=name)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/coding/times.py", line 832, in decode
    dtype = _decode_cf_datetime_dtype(data, units, calendar, self.use_cftime)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/coding/times.py", line 209, in _decode_cf_datetime_dtype
    [first_n_items(values, 1) or [0], last_item(values) or [0]]
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/core/formatting.py", line 90, in first_n_items
    return np.ravel(to_duck_array(array))[:n_desired]
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/core/pycompat.py", line 140, in to_duck_array
    return np.asarray(data)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/core/indexing.py", line 487, in __array__
    return np.asarray(self.get_duck_array(), dtype=dtype)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/core/indexing.py", line 490, in get_duck_array
    return self.array.get_duck_array()
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/core/indexing.py", line 554, in get_duck_array
    array = self.array[self.key]
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/backends/zarr.py", line 92, in __getitem__
    return array[key.tuple]
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/zarr/core.py", line 844, in __getitem__
    result = self.get_basic_selection(pure_selection, fields=fields)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/zarr/core.py", line 970, in get_basic_selection
    return self._get_basic_selection_nd(selection=selection, out=out, fields=fields)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/zarr/core.py", line 1012, in _get_basic_selection_nd
    return self._get_selection(indexer=indexer, out=out, fields=fields)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/zarr/core.py", line 1388, in _get_selection
    self._chunk_getitems(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/zarr/core.py", line 2228, in _chunk_getitems
    self._process_chunk(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/zarr/core.py", line 2141, in _process_chunk
    chunk = self._decode_chunk(cdata)
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/zarr/core.py", line 2427, in _decode_chunk
    chunk = chunk.reshape(expected_shape or self._chunks, order=self._order)
ValueError: cannot reshape array of size 1980 into shape (180,)

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "/usr/local/lib/python3.10/dist-packages/apache_beam/transforms/core.py", line 1963, in <lambda>
  File "feedstock/recipe.py", line 79, in _test_open_store
  File "feedstock/recipe.py", line 74, in _get_dataset
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/backends/api.py", line 572, in open_dataset
    backend_ds = backend.open_dataset(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/backends/zarr.py", line 1026, in open_dataset
    ds = store_entrypoint.open_dataset(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/backends/store.py", line 46, in open_dataset
    vars, attrs, coord_names = conventions.decode_cf_variables(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/conventions.py", line 447, in decode_cf_variables
    raise type(e)(f"Failed to decode variable {k!r}: {e}") from e
ValueError: Failed to decode variable 'time': cannot reshape array of size 1980 into shape (180,)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/site-packages/apache_beam/runners/worker/sdk_worker.py", line 300, in _execute
    response = task()
  File "/usr/local/lib/python3.10/site-packages/apache_beam/runners/worker/sdk_worker.py", line 375, in <lambda>
    lambda: self.create_worker().do_instruction(request), request)
  File "/usr/local/lib/python3.10/site-packages/apache_beam/runners/worker/sdk_worker.py", line 639, in do_instruction
    return getattr(self, request_type)(
  File "/usr/local/lib/python3.10/site-packages/apache_beam/runners/worker/sdk_worker.py", line 677, in process_bundle
    bundle_processor.process_bundle(instruction_id))
  File "/usr/local/lib/python3.10/site-packages/apache_beam/runners/worker/bundle_processor.py", line 1113, in process_bundle
    input_op_by_transform_id[element.transform_id].process_encoded(
  File "/usr/local/lib/python3.10/site-packages/apache_beam/runners/worker/bundle_processor.py", line 237, in process_encoded
    self.output(decoded_value)
  File "apache_beam/runners/worker/operations.py", line 570, in apache_beam.runners.worker.operations.Operation.output
  File "apache_beam/runners/worker/operations.py", line 572, in apache_beam.runners.worker.operations.Operation.output
  File "apache_beam/runners/worker/operations.py", line 263, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1526, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 636, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "apache_beam/runners/common.py", line 1621, in apache_beam.runners.common._OutputHandler.handle_process_outputs
  File "apache_beam/runners/common.py", line 1734, in apache_beam.runners.common._OutputHandler._write_value_to_tag
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1526, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 636, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "apache_beam/runners/common.py", line 1621, in apache_beam.runners.common._OutputHandler.handle_process_outputs
  File "apache_beam/runners/common.py", line 1734, in apache_beam.runners.common._OutputHandler._write_value_to_tag
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1526, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 851, in apache_beam.runners.common.PerWindowInvoker.invoke_process
  File "apache_beam/runners/common.py", line 995, in apache_beam.runners.common.PerWindowInvoker._invoke_process_per_window
  File "apache_beam/runners/common.py", line 1621, in apache_beam.runners.common._OutputHandler.handle_process_outputs
  File "apache_beam/runners/common.py", line 1734, in apache_beam.runners.common._OutputHandler._write_value_to_tag
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1526, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 636, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "apache_beam/runners/common.py", line 1621, in apache_beam.runners.common._OutputHandler.handle_process_outputs
  File "apache_beam/runners/common.py", line 1734, in apache_beam.runners.common._OutputHandler._write_value_to_tag
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1526, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 636, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "apache_beam/runners/common.py", line 1621, in apache_beam.runners.common._OutputHandler.handle_process_outputs
  File "apache_beam/runners/common.py", line 1734, in apache_beam.runners.common._OutputHandler._write_value_to_tag
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1526, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 636, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "apache_beam/runners/common.py", line 1621, in apache_beam.runners.common._OutputHandler.handle_process_outputs
  File "apache_beam/runners/common.py", line 1734, in apache_beam.runners.common._OutputHandler._write_value_to_tag
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1526, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 636, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "apache_beam/runners/common.py", line 1621, in apache_beam.runners.common._OutputHandler.handle_process_outputs
  File "apache_beam/runners/common.py", line 1734, in apache_beam.runners.common._OutputHandler._write_value_to_tag
  File "apache_beam/runners/worker/operations.py", line 266, in apache_beam.runners.worker.operations.SingletonElementConsumerSet.receive
  File "apache_beam/runners/worker/operations.py", line 953, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/worker/operations.py", line 954, in apache_beam.runners.worker.operations.DoOperation.process
  File "apache_beam/runners/common.py", line 1437, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 1547, in apache_beam.runners.common.DoFnRunner._reraise_augmented
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 637, in apache_beam.runners.common.SimpleInvoker.invoke_process
  File "/usr/local/lib/python3.10/dist-packages/apache_beam/transforms/core.py", line 1963, in <lambda>
  File "feedstock/recipe.py", line 79, in _test_open_store
  File "feedstock/recipe.py", line 74, in _get_dataset
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/backends/api.py", line 572, in open_dataset
    backend_ds = backend.open_dataset(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/backends/zarr.py", line 1026, in open_dataset
    ds = store_entrypoint.open_dataset(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/backends/store.py", line 46, in open_dataset
    vars, attrs, coord_names = conventions.decode_cf_variables(
  File "/opt/apache/beam-venv/beam-venv-worker-sdk-0-0/lib/python3.10/site-packages/xarray/conventions.py", line 447, in decode_cf_variables
    raise type(e)(f"Failed to decode variable {k!r}: {e}") from e
ValueError: Failed to decode variable 'time': cannot reshape array of size 1980 into shape (180,) [while running 'Creating CMIP6.CMIP.CCCma.CanESM5.historical.r37i1p2f1.Omon.tos.gn.v20190429|OpenURLWithFSSpec|OpenWithXarray|Preprocessor|StoreToZarr|Copy|Logging to non-QC table|TestDataset|Logging to QC table/TestDataset/Testing - Open Store-ptransform-66']
```

Seems like we need some more test cases with time arrays/enconding ase datetime and cftime?

**Comment by norlandrhagen** [comment]
Thanks for giving it a test drive @jbusecke! Yeah, seems like it. Maybe we can add another test in our Zar102 class next.

File: tests/test_rechunking.py
Comment (norlandrhagen) [reply tests/test_rechunking.py:284]: I removed it and our ~~extensive test suite~~ tests didn't complain.

**Comment by jbusecke** [comment]
Would be happy to pair on this maybe tomorrow (Wednesday)? In case that works for you.

**Comment by norlandrhagen** [comment]
That would be great! I'm a bit busy with ESIP today through Thursday, but I'm free all of Friday or next week.

**Comment by cisaacstern** [comment]
Looks like we are waiting on resolution of the error @jbusecke hit here, correct?

**Comment by jbusecke** [comment]
Ok with the current state of this PR I am getting successful writes of CMIP6 stores like this one:
`'gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-CM6-1/historical/r29i1p1f2/Omon/tos/gr1/v20200529/'`

```python
import gcsfs
fs = gcsfs.GCSFileSystem()
fs.ls('gs://cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-CM6-1/historical/r29i1p1f2/Omon/tos/gr1/v20200529/time')
```
shows that there is only one time chunk:
```
['cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-CM6-1/historical/r29i1p1f2/Omon/tos/gr1/v20200529/time/.zarray',
 'cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-CM6-1/historical/r29i1p1f2/Omon/tos/gr1/v20200529/time/.zattrs',
 'cmip6/CMIP6/CMIP/CNRM-CERFACS/CNRM-CM6-1/historical/r29i1p1f2/Omon/tos/gr1/v20200529/time/0']
```

File: pyproject.toml
Comment (cisaacstern) [inline pyproject.toml:45]: Looks like this needs to be rebased to `main`, since didn't this fix already go in?

File: pangeo_forge_recipes/transforms.py
Comment (cisaacstern) [inline pangeo_forge_recipes/transforms.py:647]:
```suggestion
```

I went back and forth on this, but IMHO I think the right path here is to not do any consolidation by default, and leave it to users to compose modularly, i.e.:

```python

recipe = (
    ...
    | StoreToZarr()
    | ConsolidateCoordinateDimensions()
    | ConsolidateMetadata()
)
```

This is more verbose for users, but it is also more explicit as to what they can expect, and it adheres to the composable ethos of Beam, and it's less API-creep for us to maintain in StoreToZarr.

Happy to discuss dissenting views here, but as of now this feels like the right path to me.

File: pangeo_forge_recipes/transforms.py
Comment (cisaacstern) [inline pangeo_forge_recipes/transforms.py:610]:
```suggestion
```

See comment below re: boolean flags vs. composition. I feel like @sbquinlan's design suggestion is the right way to go here: modularity/composibility over branching/boolean flag logic.

Note if we go this route we want to also make the following change in the body of StoreToZarr below (but GitHub won't let me make a comment on these lines because the PR does not already change them):

```diff
target_store = schema | PrepareZarrTarget(
    target=self.get_full_target(),
    target_chunks=target_chunks,
    attrs=self.attrs,
-   consolidated_metadata=self.consolidated_metadata,
+   # In StoreToZarr, always default to unconsolidated. This leaves it up to the
+   # user whether or not they want to consolidate with ConsolidateMetadata(). Also,
+   # it prevents a broken/inconsistent state that could arise from metadata being
+   # consolidated here, and then falling out of sync with coordinates if
+   # ConsolidateDimensionCoordinates() is applied to the output of StoreToZarr().
+   consolidated_metadata=False,
    encoding=self.encoding,
)

```

File: tests/test_end_to_end.py
Comment (cisaacstern) [inline tests/test_end_to_end.py:197]:
```suggestion
        unconsolidated_zarr = (
            p
            | beam.Create(pattern.items())
            | OpenWithXarray(file_type=pattern.file_type)
            | StoreToZarr(
                target_root=tmp_target,
                store_name="subpath",
                combine_dims=pattern.combine_dim_keys,
            )
        )
        if consolidate_dimension_coordinates:
            unconsolidated_zarr | ConsolidateCoordinateDimensions()

```

Assuming we go with the composable/modular design rather than the boolean flag design, then this test would need to updated to look something like above.

I'm not sure if `ConsolidateMetadata()` would also need to be piped on the end here to get the asserts below to pass.

File: tests/test_transforms.py
Comment (cisaacstern) [inline tests/test_transforms.py:314]:
```suggestion
```

Assuming we remove these flags as proposed above, they would need to be removed from this test as well.

**Review by cisaacstern — COMMENTED** [review]
Thanks so much @norlandrhagen and @jbusecke for your work on this!

I think this is very close. As I went through this, I did end up feeling rather strongly that @sbquinlan's earlier suggestion to favor modularity/composability over boolean flag configuration felt to me much more in the design spirit of Beam, and potentially a lower maintenance burden for us as developers as well. All of the "big" changes I suggest here relate to that.

Curious to know your thoughts.

File: pangeo_forge_recipes/rechunking.py
Comment (norlandrhagen) [reply pangeo_forge_recipes/rechunking.py:282]: This was the fix @jbusecke and I came up with yesterday to get around the errors reported in the cmip6 pipeline. I think when the coordinate_dims are consolidated in this PR, the metadata has to be updated after. @jbusecke  might have a better way to explain this though!

Before the consolidated_metadata PR option was introduced all of the `StoreToZarr` recipes were consolidating metadata by default.
_from the docs_ ```The__ default (consolidated=None) means write consolidated metadata and attempt to read consolidated metadata for existing stores (falling back to non-consolidated).```

File: pangeo_forge_recipes/transforms.py
Comment (norlandrhagen) [reply pangeo_forge_recipes/transforms.py:649]: Would love to chat more about this! 
I also do like the composability, but wonder about the trade off for transformations that should be done on almost every recipe, but might get overlooked by a newer recipe developer.

File: pangeo_forge_recipes/rechunking.py
Comment (cisaacstern) [reply pangeo_forge_recipes/rechunking.py:282]: Yep, that makes total sense @norlandrhagen!

Apologies the review here is a little disjointed. Down in https://github.com/pangeo-forge/pangeo-forge-recipes/pull/671#discussion_r1473536255 I mention my proposed fix for this, which is that we should _also remove_ the consolidate_metadata flag from StoreToZarr, and _always default_ `consolidated=False` in StoreToZarr.

Then it is up to the user if they want to do:

```
StoreToZarr() | ConsolidateCoordinateDimensions()
```

or

```
StoreToZarr() | ConsolidateCoordinateDimensions() | ConsolidateMetadata()
```

(☝️ seems like this is what @jbusecke would want to do)

or

```
StoreToZarr() | ConsolidateMetadata()
```

But in order to make these options possible, we need to start from an unconsolidated store emitted by `StoreToZarr()`, because as you observe here, if StoreToZarr emits a consolidated store, and coordinate dims are subsequently consolidated without then re-consolidating metadata, we end up in a broken state.

Always emitting an unconsolidated store from `StoreToZarr()` should resolve this.

Does that make sense?

Comment (norlandrhagen) [reply pangeo_forge_recipes/rechunking.py:282]: No worries!

It does fix it! However, then we are defaulting to non_consolidated coords, which seems like the wrong approach for most recipes. I think this is fine if everyone knows that they probably should add the ConsolidatedCoords transform into their recipe, but I could see it commonly slipping through the cracks. Maybe this is kind of what kind of guard rails / default assumptions we want to have 🤷 .

File: pangeo_forge_recipes/transforms.py
Comment (norlandrhagen) [reply pangeo_forge_recipes/transforms.py:586]: Good question! I don't actually know. Maybe @abarciauskas-bgse has some insight into this?

File: pangeo_forge_recipes/rechunking.py
Comment (jbusecke) [reply pangeo_forge_recipes/rechunking.py:282]: I share the concern about this requirement for explicitly having to chain consolidation stages (in a specific order) leading to errors on the user side.  

Maybe we can consider renaming `StoreToZarr`-> `StoreToZarrUnconsolidated` (or something else) and subclass that as the new `StoreToZarr` (this would change the default behavior)

alternatively make a new class `StoreToZarrConsolidated` that subclasses the current `StoreToZarr`, this would not change the default behavior but probably would need a thorough rewrite of the docs to make sure users are using consolidated storage as a default?

File: pangeo_forge_recipes/transforms.py
Comment (jbusecke) [reply pangeo_forge_recipes/transforms.py:649]: If you have a chat, I would love to attend if it fits on my schedule.

Comment (cisaacstern) [reply pangeo_forge_recipes/transforms.py:649]: Thanks for weighing in! I am not 100% sure which design is best here, and appreciate the lively discussion so we can land on the best solution. (Here and in the comment thread above.)

> done on almost every recipe, but might get overlooked by a newer recipe developer.

This is a small point, but one that just came to mind: in my experience, for smaller zarr stores, consolidation doesn't really matter / have a performance impact. Therefore, in some sense do we care if newer recipe developers (which presumably implies smaller-scale as well?) overlook this?

We absolutely should have a design discussion on this, possibly before the next coordination meeting (to expedite the process). I will try to first outline what I see as the various options in a comment on this thread to solicit async feedback.

Comment (norlandrhagen) [reply pangeo_forge_recipes/transforms.py:649]: Awesome! Happy to chat before. I won't be able to make it to the next coordination meeting, so this week or next before Friday would be ideal for me.

**Comment by norlandrhagen** [comment]
Going to hold off doing any more work on this PR until we all chat design.

File: pangeo_forge_recipes/transforms.py
Comment (abarciauskas-bgse) [reply pangeo_forge_recipes/transforms.py:586]: I don't know if there is a significant difference if we consolidate dimension coordinates in kerchunk datasets. I experienced slow performance when working with zarr stores and saw a relationship between performance (specifically for creating image tiles using rio_tiler's XarrayReader) and the number of coordinate chunks. The notebook which demonstrates this is unfortunately not a brief one (https://github.com/developmentseed/tile-benchmarking/blob/9758c732597c154d1cf7cd796b21c858c3130046/profiling/profile.ipynb) but I could streamline it if we need to do more investigation.

For kerchunk, I would guess there would be a performance impact if a coordinate's chunks are all read all at once or in multiple reads. If it's in multiple reads, it still might impact performance. Do we know which it is? Or should I investigate/run some more tests?

**Comment by norlandrhagen** [comment]
Tests are passing! @cisaacstern & @jbusecke

**Comment by norlandrhagen** [comment]
**Update to all the discussion about design above**

@cisaacstern, @jbusecke met and decided to go the route of composability instead of adding extra kwargs. This adds a bit of work onto the recipe writer, but gives them more flexibility and should reduce the maintence burden on the already very complex `StoreToZarr` transform. Thanks to @sbquinlan for pushing us down this design path.

**Pull request merged**
